core/trino-main/src/main/java/io/trino/operator/project/PageProcessor.java

```diff
@@ -331,7 +331,7 @@ private ProcessBatchResult processBatch(int batchSize)
                 else {
                     if (pageProjectWork == null) {
                         expressionProfiler.start();
-                        pageProjectWork = projection.project(session, yieldSignal, projection.getInputChannels().getInputChannels(page), positionsBatch);
+                        pageProjectWork = projection.project(session, yieldSignal, projection.getInputChannels().getLoadedInputChannels(page), positionsBatch);
```

Comment:
same here. Projection can consist of `CASE` on two columns. I think we could have a simple visitor that would detect if all columns need to be loaded unconditionally.